seq2seq/metrics/metrics.py

```diff
@@ -356,4 +356,297 @@ def eval_batch(self, outputs, targets):
+        Args:
+            outputs (list(torch.tensor)): Contains the predictions of the model. List of length max_output_length, where each element is a tensor of length batch_size
+            targets (dict): Dictionary containing the grammars
+        """
+
+        # batch_size X N variable containing the indices of the model's input,
+        # where N is the longest input
+        input_variable = targets['encoder_input']
+        output_variable = targets['decoder_output']
+        batch_size = input_variable.size(0)
+
+        # Convert to batch_size x M variable containing the indices of the model's output, where M
+        # is the longest output
+        predictions = torch.stack(outputs, dim=1).view(batch_size, -1)
+
+        # Current implementation does not allow batch-wise evaluation
+        for i_batch_element in range(batch_size):
+            # We start by counting the sequence to the total.
+            # Next we go through multiple checks for incorrectness.
+            # If all these test fail, we consider the sequence correct.
+            self.seq_total += 1
+
+            # Extract the current example and move to cpu
+            grammar = input_variable[i_batch_element, :].data.cpu().numpy()
+            prediction = predictions[i_batch_element, :].data.cpu().numpy()
+            target_output = output_variable[i_batch_element, :].data.cpu().numpy()
+
+
+            # Convert indices to strings
+            # Remove all padding from the grammar.
+            grammar = [self.input_vocab.itos[token] for token in grammar if
+                       self.input_vocab.itos[token] != self.input_pad_symbol]
+            prediction = [self.output_vocab.itos[token] for token in prediction]
+            tgt_output = [self.output_vocab.itos[token] for token in target_output if
+                          self.output_vocab.itos[token] != self.output_pad_symbol ]
+
+            # print('grammar:')
+            # print(grammar)
+            # input()
+            # print('target outputs:')
+            # print(tgt_output)
+            # print('predictions:')
+            # print(prediction)
+
+            # # The first prediction after the actual output should be EOS
+            if self.use_output_eos and prediction[-1] != self.output_eos_symbol:
+                continue
+
+            # Remove EOS (and possible padding)
+            prediction_correct_length = prediction[:-1]
+
+            # If the EOS symbol is present in the prediction, this means that the prediction was too
+            # short.
+            # Since SOS, PAD and UNK are also part of the output dictionary, these can technically
+            # also be predicted by the model, especially at the beginning of training due to random
+            # weight initialization. Since these render the output incorrect and cause an error in
+            # correct(), we check for their presence here.
+            # if  self.output_eos_symbol in prediction_correct_length or \
```

Comment:
You don't check whether the output is too short. Maybe the prediction contains two EOS symbols. One at `prediction[-1]`, but another one within `prediction_correct_length`

Reply:
short output and two eos don't affect the check for accuracy. So made prediction_correct_length = prediction .